components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt

```diff
@@ -97,16 +97,19 @@ internal object TabListReducer {
                         if (it.parentId == tabToRemove.id) it.copy(parentId = tabToRemove.parentId) else it
                     }
 
-                    val updatedSelection = if (action.selectParentIfExists && tabToRemove.parentId != null) {
-                        // The parent tab should be selected if one exists
+                    val updatedSelection: String? = if (tabToRemove.id != state.selectedTabId) {
```

Comment:
@Mugurell hm we could make this change to not select a normal tab when a private child tab is removed or vice versa. We'd definitely want unit tests though, as this is a risky change in the reducer!

However, there's definitely a bug here in Fenix too. Even if the mode changes because of removal, why is Fenix displaying the wrong tabs? Watching the video, It seems like that mode change is only partially applied.

Reply:
> However, there's definitely a bug here in Fenix too. Even if the mode changes because of removal, why is Fenix displaying the wrong tabs? Watching the video, It seems like that mode change is only partially applied.

Yes, I think so too. We're selecting a new tab that is on a different "page" in the tray, but the tabs layout at the top is unaware of that change happening.

We probably need to use `TabLayoutMediator.selectTabAtPosition` to observe the store and select the correct "page" if our selected tab type has changed.

Reply:
There is a bug in Fenix but it's not clear why and I couldn't fix it.
The `TabLayoutMediator` is set and works okay but for whatever reason upon restarting `TabsTrayFragment` (creating it again since the whole `HomeActivity` is recreated) when the mode is changed from private to normal the viewPager binds the private adapter to show the private tabs in the normal tabs position (as set by the mediator). 
Checked with the debugger when binding the adapters [here](https://github.com/mozilla-mobile/fenix/blob/8c43ccc0f6525e17cd036e4144af64c85e9d021a/app/src/main/java/org/mozilla/fenix/tabstray/TrayPagerAdapter.kt#L106) and saw that even though tabsTray#currentItem is 0 it will bind the item at position 1.

With this patch we won't go in that buggy scenario.

Reply:
Christian and I debugged this for a bit today and we learnt a few things:
 - When the theme changes, the fragment view is recreated, and we are invoking `Fragment.onViewCreated` which triggers the `TabMediator.start`. This implicitly assumes that we always start at 0 which is why the selected item is normal tabs.
 - When the `ViewPager2` is detached however, it saves it's [current state](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:viewpager2/viewpager2/src/main/java/androidx/viewpager2/widget/ViewPager2.java;l=318?q=ViewPager2) (page position), which is then restored once the fragment restores state. This explains why the tab list still shows up the private tabs, because that's where we last were before the activity recreated the view on mode change.
 
 So our options are:
 - If we want to keep the tab position as private browsing, then we need to think about how to save the state of the `TabLayout` and restore it after like the `ViewPager2`.
 - If we want to move to the page with the selected tab, this is a much smaller change by explicitly changing the page on start, with a patch like this:
 
<details>

<summary>fenix patch</summary>

 ```diff
diff --git a/app/src/main/java/org/mozilla/fenix/tabstray/TabLayoutMediator.kt b/app/src/main/java/org/mozilla/fenix/tabstray/TabLayoutMediator.kt
index 4bfeabe2b..060b1fbf9 100644
--- a/app/src/main/java/org/mozilla/fenix/tabstray/TabLayoutMediator.kt
+++ b/app/src/main/java/org/mozilla/fenix/tabstray/TabLayoutMediator.kt
@@ -5,6 +5,7 @@
 package org.mozilla.fenix.tabstray
 
 import androidx.annotation.VisibleForTesting
+import androidx.viewpager2.widget.ViewPager2
 import com.google.android.material.tabs.TabLayout
 import mozilla.components.browser.state.store.BrowserStore
 import mozilla.components.support.base.feature.LifecycleAwareFeature
@@ -21,6 +22,7 @@ import org.mozilla.fenix.utils.Do
  */
 class TabLayoutMediator(
     private val tabLayout: TabLayout,
+    private val tabPager: ViewPager2,
     interactor: TabsTrayInteractor,
     private val browsingModeManager: BrowsingModeManager,
     private val tabsTrayStore: TabsTrayStore,
@@ -55,6 +57,7 @@ class TabLayoutMediator(
 
     fun selectTabAtPosition(position: Int) {
         tabLayout.getTabAt(position)?.select()
+        tabPager.setCurrentItem(position, false)
         tabsTrayStore.dispatch(TabsTrayAction.PageSelected(Page.positionToPage(position)))
     }
 }
diff --git a/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayFragment.kt b/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayFragment.kt
index adf9d79da..df584e229 100644
--- a/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayFragment.kt
+++ b/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayFragment.kt
@@ -254,6 +254,7 @@ class TabsTrayFragment : AppCompatDialogFragment() {
         tabLayoutMediator.set(
             feature = TabLayoutMediator(
                 tabLayout = tabsTrayBinding.tabLayout,
+                tabPager = tabsTrayBinding.tabsTray,
                 interactor = tabsTrayInteractor,
                 browsingModeManager = activity.browsingModeManager,
                 tabsTrayStore = tabsTrayStore,

 ```

</details>

I think we can solve our immediate Fenix bug in that codebase without changing the behaviour in AC, and we can also discuss this patch as a separate feature request.

Reply:
Thanks @jonalmeida. Yes, I think landing ^^ the Fenix patch makes sense either way. Then mode switches will work correctly, and we can still land the A-C patch as well if we want to change the behaviour. Wdyt @Mugurell @pocmo ?

Reply:
Great! This will go in as part of https://github.com/mozilla-mobile/fenix/pull/22386

Reply:
Not quite sure about the release cycle of this project, but do you mean this will be released to the stable channel someday, which seems to be more like a "short-term solution"?

I believe there would be someone asking "why my browsing mode can be changed by closing a tab?" then, if the behavior is just as shown in this video. I'm not sure whether it would be better than a simple display issue.

Sorry if I misunderstood anything.